### ISSUE TITLE
RS0014 false positive

### ISSUE THREAD

**Issue by kentcb** [dotnet/roslyn-analyzers#1322]
#### Analyzer package

Microsoft.CodeAnalysis.FxCopAnalyzers

#### Analyzer

RS0014

#### Repro steps

```C#
someList.FirstOrDefault()
```

#### Expected behavior

No warning, or make the alternative clear. The only recourse I can think of is to implement my own `FirstOrDefault` extension against `IList<T>` (or `IImmutableList<T>`), but is that the intention?

#### Actual behavior
```
warning RS0014: Do not use Enumerable methods on indexable collections. Instead use the collection directly
```

**Comment by ivanbasov** [comment]
@kentcb, thank you for the request!  Could you please provide a complete code sample and version of code analyzers you use?

 I tried to play around  `someList.FirstOrDefault()` with the latest version of code analyzers (2.6.0-beta1-62212-01). Everything works fine.

**Comment by kentcb** [comment]
@ivanbasov I've attached the simplest possible repro.

I'm a bit confused about the version you've mentioned. AFAICT, `2.3.0-beta1` is the latest version of `Microsoft.CodeAnalysis.FxCopAnalyzers` and that's where I'm hitting the problem.

[RS0014_Repro.zip](https://github.com/dotnet/roslyn-analyzers/files/1409393/RS0014_Repro.zip)

**Comment by ivanbasov** [comment]
Thank you very much for the code sample, @kentcb!

I have verified that the code works fine with the current analyzer code 2.6.0-beta1-62212-01. This is an internal version of analyzers that can be built using current roslyn + current roslyn-analyzers repos. Nuget packages for roslyn-analyzers are expected to be updated after release of new version of IOperation which is targeted by the end of 2017.

So, I extended your code to:
```
using System.Collections.Generic;
using System.Linq;

namespace RS0014Repro
{
    public class Class1
    {
        void Use<U>(U p) { } 
        void Foo()
        {
            List<string> xs = null;
            xs.FirstOrDefault();

            IReadOnlyList<string> ys = null;
            Use(ys.FirstOrDefault());
        }
    }
}
```
the only error I see (RS0014) points to `Use(ys.FirstOrDefault());`

**Comment by kentcb** [comment]
> the only error I see (RS0014) points to Use(ys.FirstOrDefault());

Out of interest, why is that? Isn't it still the case that `FirstOrDefault` is valid in that context?

**Comment by ivanbasov** [comment]
[FirstOrDefault](https://github.com/Microsoft/referencesource/blob/master/System.Core/System/Linq/Enumerable.cs#L1056) is an universal method applicable for IEnumerable but it is not optimized for performance purposes. So, when you have an indexer like for IReadOnlyList, it would be much faster to call it directly rather than use the universal approach.

**Comment by kentcb** [comment]
Understood. So the reason it triggers against `IReadOnlyList<T>` and not `IList<T>` is because `FirstOrDefault` is special-cased for `IList<T>`? In that case, it will still trigger for `IImmutableList<T>`, which was my original use-case. Hence, I am back to my original query about what the recourse should be. I think the analyzer should provide more info in this respect. For example, I could just write my own `FirstOrDefault` extension against `IImmutableList<T>` (or `IReadOnlyList<T>`), but it would be good if that was made clear by the analyzer.

**Comment by ivanbasov** [comment]
Thank you, @kentcb! Yes, IList is a special case. Your understanding is correct. It matches with the [comment ](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.cs#L91) to the analyzer. And, yes, we can consider to extend it as mentioned in the comment.
